sql: track uniqueness of results in ordering information

RaduBerinde opened the pull request:
Consider a very basic example: on `TABLE kv (k INT PRIMARY KEY, v INT)`
we do a `SELECT * from kv ORDER BY k, v`.

Before this change we would sort the results but that's not necessary: the
data is already sorted by `k` and there are no duplicate values of `k` so the
extra refinement by `v` can be ignored.

In this change we fix cases like this by adding a `unique` flag to
`orderingInfo`; if set it indicates that we cannot have duplicate tuples for the
columns in the ordering (and thus any further refinement of results by other
columns is a no-op).

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4798)

<!-- Reviewable:end -->

petermattis commented:
LGTM

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/testdata/order_by, line 207 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4798#-KBxKXILWMBP5s__-975:-KBxKXIMjNxJwgd2sy9Q:1945647986) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a74b649857b61ce34d16171c8f159340f0d9f54/sql/testdata/order_by#L207)):*
I found this comment a bit confusing. Perhaps: `the non-unique index ba will include column c (to make the keys unique) so we won't need to sort.`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4798#-:-KBxLDQNeiRzG27H5_X1:676834233)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[sql/testdata/order_by, line 207 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4798#-KBxKXILWMBP5s__-975:-KBxORThjQIKurJVt1Up:1138235560) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a74b649857b61ce34d16171c8f159340f0d9f54/sql/testdata/order_by#L207)):*
Fixed and added another test (this particular one would have worked without this change).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4798)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: track uniqueness of results in ordering information

Consider a very basic example: on `TABLE kv (k INT PRIMARY KEY, v INT)`
we do a `SELECT * from kv ORDER BY k, v`.

Before this change we would sort the results but that's not necessary: the
data is already sorted by `k` and there are no duplicate values of `k` so the
extra refinement by `v` can be ignored.

In this change we fix cases like this by adding a `unique` flag to
`orderingInfo`; if set it indicates that we cannot have duplicate tuples for the
columns in the ordering (and thus any further refinement of results by other
columns is a no-op).
